### PR TITLE
Make Topology a singleton (and fix the topology representation for non-NUMA systems)

### PR DESCRIPTION
**Pull request by lanice**
Originally I had this as part of #969, but this is a change that stands on it's own, and since it is something that affects a big part of Hyrise, I wanted to not just sneak it in with the other PR.

This PR makes the Topology class a singleton, that is initialized with the default topology settings when first called `Topology::current()`. From then, if an application (or test/benchmark for that matter) needs a different topology (e.g. a fake topology with specific nodes/cores), it can create one via `Topology::current().create_fake_numa_topology()` and similar methods.

The reason this came up was, because when I worked on the MPSM Join, I suddenly had to create different branches for NUMA / non-NUMA systems JUST to create the correct Topology and get the node information. And once more and more operators become NUMA-aware, instead of everyone doing it's own Topology-creation-handling, it rather should just be a call to `Topology::current()`, to know on what system we are running.

This also adds a proper topology representation for non-NUMA systems, which is 1 node and X cores usually. When you activated the Scheduler before (on non-NUMA systems), you had a weird fake NUMA topology, now you have your actual system topology.

Side effect of this is, that applications that used the scheduler (like `BenchmarkRunner` or `Console`) now actually use the correct topology on non-NUMA systems (1 node, X cores), when before they used a fake NUMA topology (X nodes, 1 core each), which doesn't really make sense.

I fiddled quite a bit with the implementation, and singletons may not be the best thing to do, I know. So if anyone has suggestions for improvements or changes, don't hesitate.

### CODE REVIEW

File: src/test/scheduler/scheduler_test.cpp [hunk 1]
```diff
@@ -140,7 +142,8 @@ TEST_F(SchedulerTest, LinearDependenciesWithScheduler) {
 }
 
 TEST_F(SchedulerTest, MultipleDependenciesWithScheduler) {
-  CurrentScheduler::set(std::make_shared<NodeQueueScheduler>(Topology::create_fake_numa_topology(8, 4)));
+  Topology::current().create_fake_numa_topology(8, 4);
```
Comment (mrzzzrm) [inline src/test/scheduler/scheduler_test.cpp]: I find that a bit unintuitive to read, I would expect a `create_*` to return the new topology, not to reset the current topology to it. `Topology::current().init_fake_numa_topology(8, 4);`, maybe?

File: src/lib/scheduler/topology.hpp [hunk 2]
```diff
@@ -29,15 +29,12 @@ struct TopologyNode final {
...
-   */
-  static std::shared_ptr<Topology> create_fake_numa_topology(uint32_t max_num_workers = 0,
-                                                             uint32_t workers_per_node = 1);
-  static std::shared_ptr<Topology> create_numa_topology(uint32_t max_num_cores = 0);
+  static Topology& current();
 
-  Topology(std::vector<TopologyNode>&& nodes, size_t num_cpus) : _nodes(std::move(nodes)), _num_cpus(num_cpus) {}
+  void create_default_topology();
+  void create_numa_topology(uint32_t max_num_cores = 0);
+  void create_nonnuma_topology(uint32_t max_num_cores = 0);
```
Comment (mrzzzrm) [inline src/lib/scheduler/topology.hpp]: Really minor, but I'd prefer `*_non_numa_*` since there is no such word "nonnuma"

**Comment by lanice** [comment]
@mrzzzrm Yeah, that was my biggest concern too about this whole thing. It is a bit similar to our `CurrentScheduler` setup I think. It does finish all tasks before setting a new Scheduler, but still you can not be 100% sure that if you set a specific Scheduler, that at any later point there is not another Scheduler set from somewhere else. Except of course we only have one Scheduler at the moment.

But similar to the Scheduler thing, the concept here should be that IF you need a custom Topology, you should set it at the beginning of your program, and do any multi threaded logic afterwards without touching the Topology again.

The other option I thought of was to add some kind of security mechanism in `Topology::current()` that validates if the Topology is still the same, for example:
The `Topology::init_*` methods return a token instead of `void`.
`Topology::current()` takes a token as optional argument.
If the passed token does not match the current Topology, do something (Assert, panic, I don't know).

Just a thought, don't know if that's viable at all, or we should just make it a "use with care" thing...

File: src/test/scheduler/scheduler_test.cpp (hunk 1 above)
Comment (lanice) [reply src/test/scheduler/scheduler_test.cpp]: I like `init_*` :+1:

**Comment by mrzzzrm** [comment]
I think, while not perfect, "use with care" is the most practical solution.
Tests can change the topology from case to case, and all other applications need to be careful.

**Pull request merged**